### DISCUSSION TITLE
Structure based navigation and manipulations

### DISCUSSION THREAD

**Discussion by mmontone** [neomacs-project/neomacs#74] [Ideas]
Have commands for navigation and manipulation based on structure.

A possible list:
* Navigation:
   - next-sibling: Move focus to next sibling.
   - prev-sibling: Move focus to previous sibling.
   - move-to-parent: Move focus parent node.
   - move-to-child: Move focus to first child node.
* Manipulations:
   - delete-focused: Deleted currently focused node.
   - kill-focused: Delete and copy currently focused node.
   - swap-with-next: Swap current node with next sibling. 
   - swap-with-prev: Swank current node with prev sibling.

The good thing is that these will translate to every mode, not only Lisp mode.

I would like to implement, but I'm finding it a bit difficult for now because I don't fully understand Neomacs model yet (marks/nodes/positions/elements).

Thoughts?

**Comment by kchanqvq** [comment]
Looks good! Thanks for sharing your thoughts!

The motion commands are mostly mode-agnostic so far. 

- next-sibling: we have forward-element, forward-element-end
- prev-sibling: we have backward-element
- move-to-parent, move-to-child: don't have them yet, but should be straightforward to add

I'm able to make them mode-agnostic largely because of `selectable-p-aux`, which allows modes to define positions that make sense for them. Caveat: one could argue that the above set is not "fully-structural" -- what if a mode has a different idea of abstract node from concrete DOM elements? In practice I found them mostly sufficient, because their behavior can be customized by banning unwanted positions using `selectable-p-aux`.

Manipulations:

I don't have much them yet, because I once had a hard time figuring out how to make them mode-agnostic (and not breaking invariants that modes respect, while we don't have a schema system yet!). But now I think maybe we can just do the same thing as how I did for motion -- use DOM element as the basic unit. Note that the commands you propose don't change sibling level (i.e. not moving up or down nodes) -- that means if the imaginary schema system only supports "can contains" relation, then these command won't ever break the schema. We still need more thoughts if we want to add commands that move up or down nodes, though.

- delete-focused, kill-focused: we have cut-element, copy-element
- swap-with-prev, swap-with-next: don't have them yet, but doable
- raise-node, descent-node...: how?

**Reply by mmontone** [reply]
I was not aware of the `*-element` family of commands. I'll take a look.

**Reply by mmontone** [reply]
> * move-to-parent, move-to-child: don't have them yet, but should be straightforward to add

there's backward-up-node

**Reply by kchanqvq** [reply]
> there's backward-up-node

Yes, good find!

**Reply by mmontone** [reply]
I'm trying to implement the swapping commands. I've got this atm:
```lisp
(defun node-focused (marker-or-pos)
  "Return the node at MARKER-OR-POS."
  (let ((pos (resolve-marker marker-or-pos)))
    (ematch pos
      ((element) pos)
      ((end-pos node) node)
      ((text-pos node) (parent node))
      (nil))))

(define-command swap-next-sibling (&optional (marker (focus)))
  (let* ((node (node-focused marker))
	 (parent (parent node))
	 (next-sibling (next-sibling node)))
    (when next-sibling
      (delete-node next-sibling)
      (insert-before parent next-sibling node))))
```
But it is not working, corrupts the document. Do you know what is wrong? (you can assume `node-focused` is working, the problem is in `swap-next-sibling` node manipulations).

**Reply by mmontone** [reply]
I'm probably misusing the low-level dom operations `insert-before`.
These high-level macro-like kinda work:
```lisp
(define-command swap-next-sibling ()
  (forward-element)
  (cut-element)
  (backward-element)
  (paste))

(define-command swap-prev-sibling ()
  (cut-element)
  (backward-element)
  (paste)
  (backward-element))
```

**Reply by kchanqvq** [reply]
> I'm probably misusing the low-level dom operations `insert-before`.

Yes, using these on any "live" nodes guarantees disaster. I thought about making them internal but they are useful for non-live nodes, so I'm not sure what to do about it yet.

Try the following
```
(define-command swap-next-element (&optional (pos (focus)))
  (let* ((element (or (pos-up-ensure pos #'element-p)
                      (error 'top-of-subtree)))
         (dst (or (pos-right (pos-right element))
                  (error 'end-of-subtree))))
    (move-nodes element (pos-right element) dst)))

(define-command swap-previous-element (&optional (pos (focus)))
  (let* ((element (or (pos-up-ensure pos #'element-p)
                      (error 'top-of-subtree)))
         (dst (or (pos-left element) (error 'end-of-subtree))))
    (move-nodes element (pos-right element) dst)))
```

**Reply by kchanqvq** [reply]
I played with these a little while and I think maybe it should goes up until some parent has a next/previous sibling? That is more DWIM feel, i.e. `((((^foo))) bar)` -swap-next-element-> `(bar (((^foo))))` (`^` is the focus position).

**Reply by mmontone** [reply]
They work fine, mostly. IMO it is fine that they work with a fixed level. Like mine, they work well with top-level html, but no so much in lisp mode, because there are `br` elements involved. I'm having a look at that ...

**Reply by mmontone** [reply]
These implement the kind of manipulation I would expect. They even work in Lisp mode with newlines in-between:

```lisp
(defun node-focused (&optional (marker-or-pos (focus)))
  (or (pos-up-ensure marker-or-pos #'element-p)
      (error 'top-of-subtree)))

(defun is-br-p (node)
  (tag-name-p node "br"))

(defun next-sibling-if (predicate element)
  (let ((next-sibling (next-sibling element)))
    (when (not next-sibling)
      (return-from next-sibling-if nil))
    (if (funcall predicate next-sibling)
        next-sibling
        (next-sibling-if predicate next-sibling))))

(defun prev-sibling-if (predicate element)
  (let ((prev-sibling (previous-sibling element)))
    (when (not prev-sibling)
      (return-from prev-sibling-if nil))
    (if (funcall predicate prev-sibling)
        prev-sibling
        (prev-sibling-if predicate prev-sibling))))

(defun swap-nodes (node1 node2)
  (let ((swap-node (make-instance 'element :tag-name "swap"))) ;; intermediate node for doing the swap
    (replace-node node1 swap-node)
    (replace-node node2 node1)
    (replace-node swap-node node2)))

(define-command swap-next-sibling (&optional (marker (focus)))
  (let* ((node-focused (node-focused marker))
         (next-sibling (next-sibling-if (lambda (sibling)
                                          (and (element-p sibling)
                                               (not (is-br-p sibling))))
                                        node-focused)))
    (when (not next-sibling)
      (return-from swap-next-sibling))
    (swap-nodes node-focused next-sibling)
    (setf (pos marker) (pos-up (end-pos node-focused)))
    ))

(define-command swap-prev-sibling (&optional (marker (focus)))
  (let* ((node-focused (node-focused marker))
         (prev-sibling (prev-sibling-if (lambda (sibling)
                                          (and (element-p sibling)
                                               (not (is-br-p sibling))))
                                        node-focused)))
    (when (not prev-sibling)
      (return-from swap-prev-sibling))
    (swap-nodes node-focused prev-sibling)
    (setf (pos marker) (pos-up (end-pos node-focused)))))
```

**Reply by kchanqvq** [reply]
Good point about new line! I fixed them, take a look at `swap-next-element` and `swap-previous-element` in `edit.lisp`. Your implementation reinvented a few functions that were already there -- `new-line-node-p, pos-right-until, pos-left-until`. Also using `move-nodes` directly is probably more suited to this use case. `replace-node` is implemented using `move-nodes` under the hood.

These command should have key bindings, any suggestions?

Do you have any objections against the DWIM-style behavior? If not I'll probably add them tomorrow, because I feel like they're more convenient.

**Reply by mmontone** [reply]
They seem to work like mine, but, some strange thing: sometimes the commands do nothing. I have to repeat them. Don't you see same problem? There's no such problem in my implementation.

About key bindings, I'm binding to "shift" + arrow key for swaps, and "ctrl" + arrow key for navigation, but I don't see those as definitive. 

I'm not sure about the DWIM. I like the current behaviour, but I would have to test the DWIM to tell. Why not have both sets of commands, DWIM and the current? Then user decides by setting up keybindings.

Btw, these are some customizations I use:

```
(define-command eval-inspect-last-expression :mode lisp-mode
  (&optional (marker (focus)))
  "Evaluate the last expression and inspect the result."
  (let* ((*package* (current-package marker))
         (result (eval (node-to-sexp (last-expression (pos marker))))))
    (inspect-object result)))

;; for development/debugging:
(define-command inspect-focus ()
  "Inspect the focused node."
  (inspect-object (focus)))

(define-command what-focus ()
  "Display the focused node."
  (message (prin1-to-string (focus))))
```
And keybindings, for direct undo (don't always want to activate undo tree), and forward deletion to `del` key:
```
(set-key *global-keymap* "C-z" 'undo-command)
(set-key *global-keymap* "delete" 'forward-delete)

(set-key *global-keymap* "C-arrow-right" 'forward-element)
(set-key *global-keymap* "C-arrow-left" 'backward-element)
(set-key *global-keymap* "C-arrow-up" 'backward-up-node)
(set-key *global-keymap* "S-arrow-right" 'swap-next-element)
(set-key *global-keymap* "S-arrow-left" 'swap-previous-element)
```

**Reply by mmontone** [reply]
> I'm not sure about the DWIM. I like the current behaviour, but I would have to test the DWIM to tell. Why not have both sets of commands, DWIM and the current? Then user decides by setting up keybindings.

Or introduce a `*dwim-navigation*` boolean variable that those commands use to toggle DWIM.

**Reply by mmontone** [reply]
> Good point about new line! I fixed them, take a look at `swap-next-element` and `swap-previous-element` in `edit.lisp`.

They don't work with newline like my commands.

**Reply by kchanqvq** [reply]
Indeed, I played with your command and the newline handling make more sense, thanks! I improved mine to adopt your behavior.

> sometimes the commands do nothing.

Does the problem persist? Can you provide a test case?

Thanks for sharing your customizations btw!

**Reply by mmontone** [reply]
> > sometimes the commands do nothing.
> 
> Does the problem persist? Can you provide a test case?

The problem is fixed. One last thing that my implementation does, is to move the focus . So if I have:
```
^foo
bar
baz
```
and swap next:
```
bar
^foo
baz
```
and again:
```
bar
baz
^foo
```

The focus is kept in foo, like I'm moving it, structurally. Both for next and for previous swaps.

**Reply by kchanqvq** [reply]
My implementation should does that as well. Did you find a case where it fails?

**Reply by mmontone** [reply]
I always fails :) Focus never moves, is maintained at same position after the swap.

**Reply by mmontone** [reply]
Auch. Wait. My implementations are overwriting yours `swap-nodes` :/

**Reply by mmontone** [reply]
They work perfectly actually :) It was a name clash with my custom functions.
